src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey_slh_dsa.c

```diff
@@ -0,0 +1,280 @@
+                                    uint8_t* destination, int32_t destinationLen)
+{
+
+    assert(pkey);
+    assert(msgLen >= 0);
+    assert(contextLen >= 0);
+    assert(destination);
+    assert(destinationLen >= 7856 /* SLH-DSA-SHA2-128s/SLH-DSA-SHAKE-128s signature size */);
+
+#if defined(NEED_OPENSSL_3_0) && HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return -1;
+    }
+
+    ERR_clear_error();
+
+    EVP_PKEY_CTX* ctx = NULL;
+
+    int ret = -1;
+
+    ctx = EvpPKeyCtxCreateFromPKey(pkey, extraHandle);
+    if (!ctx)
+    {
+        goto done;
+    }
+
+    OSSL_PARAM contextParams[] =
+    {
+        OSSL_PARAM_construct_end(),
+        OSSL_PARAM_construct_end(),
+    };
+
+    if (context)
+    {
+        contextParams[0] = OSSL_PARAM_construct_octet_string(OSSL_SIGNATURE_PARAM_CONTEXT_STRING, (void*)context, Int32ToSizeT(contextLen));
+    }
+
+    if (EVP_PKEY_sign_message_init(ctx, NULL, contextParams) <= 0)
+    {
+        goto done;
+    }
+
+    size_t dstLen = Int32ToSizeT(destinationLen);
+    if (EVP_PKEY_sign(ctx, destination, &dstLen, msg, Int32ToSizeT(msgLen)) == 1)
+    {
+        if (dstLen != Int32ToSizeT(destinationLen))
+        {
+            assert(false); // length mismatch
+            goto done;
+        }
+
+        ret = 1;
+    }
+    else
+    {
+        ret = 0;
+    }
+
+done:
+    if (ctx != NULL) EVP_PKEY_CTX_free(ctx);
+    return ret;
+#else
+    (void)pkey;
+    (void)extraHandle;
+    (void)msg;
+    (void)msgLen;
+    (void)context;
+    (void)contextLen;
+    (void)destination;
+    (void)destinationLen;
+    return -1;
+#endif
+}
+
+int32_t CryptoNative_SlhDsaVerifyPure(EVP_PKEY *pkey,
+                                      void* extraHandle,
+                                      uint8_t* msg, int32_t msgLen,
+                                      uint8_t* context, int32_t contextLen,
+                                      uint8_t* sig, int32_t sigLen)
+{
+    assert(pkey);
+    assert(msgLen >= 0);
+    assert(sig);
+    assert(sigLen >= 7856 /* SLH-DSA-SHA2-128s/SLH-DSA-SHAKE-128s signature size */);
+    assert(contextLen >= 0);
+
+#if defined(NEED_OPENSSL_3_0) && HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return -1;
+    }
+
+    ERR_clear_error();
+
+    EVP_PKEY_CTX* ctx = NULL;
+
+    int ret = -1;
+
+    ctx = EvpPKeyCtxCreateFromPKey(pkey, extraHandle);
+    if (!ctx)
+    {
+        goto done;
+    }
+
+    OSSL_PARAM contextParams[] =
+    {
+        OSSL_PARAM_construct_end(),
+        OSSL_PARAM_construct_end(),
+    };
+
+    if (context)
+    {
+        contextParams[0] = OSSL_PARAM_construct_octet_string(OSSL_SIGNATURE_PARAM_CONTEXT_STRING, (void*)context, Int32ToSizeT(contextLen));
```

Comment:
I think this entire method is identical to ML-DSA minus asserts, we might want to consider unifying the core part (and same for verify)

Reply:
I'm open to squeezing it down to shared things, but as a followup.  Naming and shaping being the hard parts, of course :smile:.  The earlier attempt didn't feel right, and just having an entirely separate function felt/feels like the best short term answer, rather than ratholing on the shape and blocking other progress.